rafthttp/stream.go

```diff
@@ -0,0 +1,279 @@
+package rafthttp
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+var (
+	streamBufSize = 4096
+	errHeartbeat  = errors.New("the message is heartbeat")
+)
+
+type encoder interface {
+	// isEllegible returns whether the message is ellegible to be encoded.
+	// You must ensure the message is ellegible before calling encodeTo.
+	isEllegible(m raftpb.Message) bool
+	// loadMetadata loads the metadata from the given info. It updates
+	// the intenal status and attributes of the encoder.
+	loadMetadata(info http.Header)
+	// encodeTo encodes the given message to an output stream. The
+	// wire-format message must not start with big-endian uint64 zero, which
+	// is reserved as heartbeat.
+	encodeTo(w io.Writer, m raftpb.Message) error
+}
+
+type decoder interface {
+	// needReset returns whether it needs to reset in-using internal
+	// attributes.
+	needReset() bool
+	// saveMetadata saves the latest metadata into the given info. From now
+	// on, it uses the latest metadata as internal attributes.
+	saveMetadata(info http.Header)
+	// decodeFrom decodes the message from an input stream. When it reads
+	// the data that starts with big-endian uint64 zero, it must return
+	// errHeartbeat.
+	decodeFrom(r io.Reader) (raftpb.Message, error)
+}
+
+type WriteFlushCloser interface {
+	io.WriteCloser
+	http.Flusher
+}
+
+type outgoingConn struct {
+	WriteFlushCloser
+	info http.Header
+}
+
+type streamWriter struct {
+	enc encoder
+
+	mu      sync.Mutex // guard field working
+	working bool
+	msgc    chan raftpb.Message
+	connc   chan outgoingConn
+	stopc   chan struct{}
+	done    chan struct{}
+}
+
+func startStreamWriter(enc encoder) *streamWriter {
+	w := &streamWriter{
+		enc:   enc,
+		msgc:  make(chan raftpb.Message, streamBufSize),
+		connc: make(chan outgoingConn),
+		stopc: make(chan struct{}),
+		done:  make(chan struct{}),
+	}
+	go w.run()
+	return w
+}
+
+func (cw *streamWriter) run() {
+	// msgc is available iff w is valid
+	var msgc chan raftpb.Message
+	var heartbeatc <-chan time.Time
+	var w WriteFlushCloser
+	tickc := time.Tick(ConnReadTimeout / 3)
+	reset := func(nw WriteFlushCloser) {
+		if w != nil {
+			w.Close()
+		}
+		if nw != nil {
+			heartbeatc, msgc = tickc, cw.msgc
+		} else {
+			heartbeatc, msgc = nil, nil
+		}
+		w = nw
+		cw.mu.Lock()
+		cw.working = nw != nil
+		cw.mu.Unlock()
+	}
+	for {
+		select {
+		case <-heartbeatc:
+			// send empty message as heartbeat
+			if err := binary.Write(w, binary.BigEndian, uint64(0)); err != nil {
+				log.Printf("rafthttp: encountered error heartbeating to connection: %v", err)
+				reset(nil)
+				continue
+			}
+			w.Flush()
+		case m := <-msgc:
+			if !cw.enc.isEllegible(m) {
+				log.Printf("rafthttp: silently dropped message %s because it is not ellegible", m.Type)
```

Comment:
It can see that the message is dropped intentionally by rafthttp.
But if that happens, that means they are messages for previous terms, so the message is quite useless. I will delete the log.